### PR TITLE
Introduce helper for handling of post-handshake handshake messages in TLS <=1.2

### PR DESCRIPTION
**Pull request by hanno-arm**
Handling the receipt of a handshake record after the initial handshake requires non-trivial logic depending on the protocol version and the endpoint. This logic is currently embedded in `mbedtls_ssl_read()`.

With the introduction of support for [D]TLS 1.3, the logic will become even more complex, since [D]TLS 1.3 drops support for renegotiation -- which in [D]TLS 1.2 is the main purpose of post-handshake handshake messages -- but instead introduces numerous other post-handshake handshake messages.

In order to pave the way for those changes, this PR improves readability and maintainability of `mbedtls_ssl_read()` by moving the TLS <=1.2 logic for handling post-handshake handshake messages into a separate helper function 
 `ssl_handle_hs_message_post_handshake()`.

The logic of the code should be entirely unchanged.

### CODE REVIEW

File: library/ssl_msg.c [hunk 1]
```diff
@@ -5527,135 +5652,15 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
...
-                {
-                    MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
-                    return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
-                }
+                MBEDTLS_SSL_DEBUG_RET( 1, "ssl_handle_hs_message_post_handshake",
+                                          ret );
+                return( ret );
             }
 
-            /* At this point, we don't know whether the renegotiation has been
```
Comment (gilles-peskine-arm) [inline library/ssl_msg.c:5641]: Why did you remove this comment? It still seems useful to me.

File: library/ssl_msg.c [hunk 2]
```diff
@@ -5429,6 +5429,131 @@ static int ssl_check_ctr_renegotiate( mbedtls_ssl_context *ssl )
...
+ * may only be sent for the purpose of initiating renegotiations.
+ *
+ * This function is introduced as a separate helper since the handling
+ * of post-handshake handshake messages changes significantly in TLS 1.3,
+ * and having a helper function allows to distinguish between TLS <= 1.2 and
+ * TLS 1.3 in the future without bloating the logic of mbedtls_ssl_read().
+ */
+static int ssl_handle_hs_message_post_handshake( mbedtls_ssl_context *ssl )
+{
+    int ret;
```
Comment (gilles-peskine-arm) [inline library/ssl_msg.c:5443]:
```suggestion
    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I replayed part of the change locally then looked at the remaining diff with this PR. Looks good to me apart from two minor issues.

**Review by paul-elliott-arm — COMMENTED** [review]
Generally agreed with Gilles. Looks like a good refactor, other than that.

File: library/ssl_msg.c (hunk 2 above)
Comment (hanno-arm) [reply library/ssl_msg.c:5443]: Fixing.

File: library/ssl_msg.c (hunk 1 above)
Comment (hanno-arm) [reply library/ssl_msg.c:5641]: Ok, I'm re-introducing it.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm @paul-elliott-arm Thank you for your review, and apologies for the long silence. I've addressed your comments. Could you please re-review?

**Comment by hanno-arm** [comment]
@gilles-peskine-arm @paul-elliott-arm I rebased to resolve a merge conflict stemming from the removal of SSLv3 renegotiation rejection which was part of the code being moved into `ssl_handle_hs_message_post_handshake()`.

Could you please re-review?

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by paul-elliott-arm — APPROVED** [review]
LGTM
